src/Lorentz/Contracts/GameDAO.hs

```diff
@@ -0,0 +1,173 @@
+
+instance TypeHasDoc ItemChange where
+  typeDocMdDescription = "Describe the update of a specific item."
+
+instance HasAnnotation ItemChange where
+  annOptions = DAO.baseDaoAnnOptions
+
+data HeroChange = HeroChange
+  { hcHeroName :: MText
+  , hcChangelogs :: [MText]
+  }
+  deriving stock (Generic)
+  deriving anyclass (IsoValue)
+
+instance TypeHasDoc HeroChange where
+  typeDocMdDescription = "Describe the update of a specific hero."
+
+instance HasAnnotation HeroChange where
+  annOptions = DAO.baseDaoAnnOptions
+
+type NewContents = [NewContent]
+
+data NewContent = NewContent
+  { ncContentName :: MText
+  , ncContentDescription :: [MText]
+  }
+  deriving stock (Generic)
+  deriving anyclass (IsoValue)
+
+instance TypeHasDoc NewContent where
+  typeDocMdDescription = "Describe a new content. The content should have a name and detail description."
+
+instance HasAnnotation NewContent where
+  annOptions = DAO.baseDaoAnnOptions
+
+instance TypeHasDoc GameDaoProposalMetadata where
+  typeDocMdDescription = "GameDAO's metadata. This fields affect how proposal got accepted and how many tokens will be frozen"
+
+----------------------------------------------------------------------------
+
+type Parameter = DAO.Parameter GameDaoProposalMetadata
+type Storage = DAO.Storage GameDaoProposalMetadata
+
+emptyStorage :: Storage
+emptyStorage = DAO.emptyStorage
+  { DAO.sVotingPeriod = 60 * 60 * 24 * 7 -- 7 days
+  , DAO.sQuorumThreshold = 100 -- All player votes which show that this proposals count
+  }
+
+-- | For "balance change", the proposer needs to freeze 10 tokens
+-- For "new content", the proposal needs to freeze 50 tokens
+gameDaoProposalCheck
+  :: forall s. (DAO.ProposeParams GameDaoProposalMetadata) : s
+  :-> Bool : s
+gameDaoProposalCheck = do
+  getFieldNamed #ppFrozenToken; swap
+  toField #ppProposalMetadata
+  toField #pmProposalType
+  stackType @(ProposalType : "ppFrozenToken" :! Natural : s)
+  caseT
+    ( #cBalanceType /-> do drop; push (10 :: Natural)
+    , #cNewType /-> do drop; push (50 :: Natural)
+    )
+  toNamed #requireValue
+
+  if #requireValue <=. #ppFrozenToken then
+    push True
+  else
+    push False
+
+-- | In case of a proposal got rejected, the proposer loses half of the tokens.
+gameDaoRejectedProposalReturnValue
+  :: forall s. (DAO.Proposal GameDaoProposalMetadata) : s
+  :-> ("slash_amount" :! Natural) : s
+gameDaoRejectedProposalReturnValue = do
+  toField #pProposerFrozenToken
+  push (2 :: Natural)
+  swap
+  ediv
+  ifSome car $
+    push (0 :: Natural)
+  toNamed #slash_amount
+
+config :: DAO.Config GameDaoProposalMetadata
+config = DAO.defaultConfig
+  { DAO.cDaoName = "Game DAO"
+  , DAO.cDaoDescription = [md|A simple DAO for a Moba-like game.|]
+  , DAO.cProposalCheck = gameDaoProposalCheck
+  , DAO.cRejectedProposalReturnValue = gameDaoRejectedProposalReturnValue
+  , DAO.cDecisionLambda = do
+      -- TODO: Come up with interesting decision lambda
+      -- Can it call http endpoint?
```

Comment:
:joy_cat: :joy_cat: :joy_cat: 

That would be really interesting if it could

Reply:
Maybe instead of `http` endpoint we could have a separate contract that stores game logs, or maybe top 5 proposals? and in `decisionLambda` we can call that contract to include the recent events.

Reply:
Calling another contract sounds interesting indeed.